dipy/align/vector_fields.pyx

```diff
@@ -919,13 +923,91 @@ def compose_vector_fields_2d(floating[:, :, :] d1, floating[:, :, :] d2,
     return np.asarray(comp), np.asarray(stats)
 
 
+cdef void _compose_vector_fields_3d_fun(floating[:, :, :, :] d1,
+                                        floating[:, :, :, :] d2,
+                                        double[:, :] premult_index,
+                                        double[:, :] premult_disp,
+                                        double t, floating[:, :, :, :] comp,
+                                        cnp.npy_intp k, int *cnt_ptr,
+                                        double *maxNorm_ptr,
+                                        double *meanNorm_ptr,
+                                        double *stdNorm_ptr) nogil:
+    cdef:
+        cnp.npy_intp nr1 = d1.shape[1]
+        cnp.npy_intp nc1 = d1.shape[2]
+        int inside
+        double nn
+        cnp.npy_intp i, j
+        double di, dj, dk, dii, djj, dkk, diii, djjj, dkkk
+
+    cnt_ptr[k] = 0
+    maxNorm_ptr[k] = 0
+    meanNorm_ptr[k] = 0
+    stdNorm_ptr[k] = 0
+
+    for i in range(nr1):
+        for j in range(nc1):
+
+            # This is the only place we access d1[k, i, j]
```

Comment:
introducing one more for cycle instead repetitive using triple commands would make the code much more readable